File: src/geometry/region.zig
const std = @import("std");
const insertion = std.sort.insertion;
const powi = std.math.powi;

const geometry = @import("geometry.zig");

/// Defines an extended region around a block that can optionally include ghost nodes. In 2D, a region subdivision looks like
///
///     ---------------
///     | NW | N | NE |
///     ---------------
///     | W  |   | E  |
///     ---------------
///     | SW | S | SE |
///     ---------------
///
/// And allows one to write algorithms that traverse the edges and corners in an ordered way and iterate over the
/// cells within a given region. Most functions and iterators accept an extent parameter `E` which determines the
/// additional width of these buffer regions and a `block` parameter which describes the size of the central block.
pub fn Region(comptime N: usize) type {
    return struct {
        sides: [N]Side,

        const Self = @This();
        const AxisMask = @import("axis.zig").AxisMask(N);
        const FaceIndex = @import("axis.zig").FaceIndex(N);
        const IndexBox = @import("box.zig").IndexBox(N);
        const IndexSpace = @import("index.zig").IndexSpace(N);

        pub const count = blk: {
            var result: usize = 1;

            for (0..N) |_| {
                result *= 3;
            }

            break :blk result;
        };

        // Which side the region is on for each axis.
        pub const Side = enum(u2) {
            left = 0,
            middle = 1,
            right = 2,
        };

        /// Computes the opposing region. For instance E -> W and NW -> SE.
        pub fn reversed(self: @This()) @This() {
            var result: Self = undefined;

            for (0..N) |axis| {
                result.sides[axis] = switch (self.sides[axis]) {
                    .left => .right,
                    .middle => .middle,
                    .right => .left,
                };
            }

            return result;
        }

        /// Converts a region into a linear index into an array.
        pub fn toLinear(self: @This()) usize {
            const rspace = IndexSpace.fromSize([1]usize{3} ** N);

            var cart: [N]usize = undefined;

            for (0..N) |axis| {
                cart[axis] = @intFromEnum(self.sides[axis]);
            }

            return rspace.linearFromCartesian(cart);
        }

        /// How many steps of adjacency to reach the middle region.
        pub fn adjacency(self: Self) usize {
            var res: usize = 0;

            for (0..N) |i| {
                if (self.sides[i] != .middle) {
                    res += 1;
                }
            }

            return res;
        }

        pub fn adjacentFaces(comptime self: Self) [self.adjacency()]FaceIndex {
            comptime {
                var result: [self.adjacency()]FaceIndex = undefined;
                var cursor: usize = 0;

                for (0..N) |axis| {
                    if (self.sides[axis] == .middle) {
                        // We need not do anything
                        continue;
                    }

                    const face = FaceIndex{
                        .side = self.sides[axis] == .right,
                        .axis = axis,
                    };

                    result[cursor] = face;

                    cursor += 1;
                }
            }
        }

        pub fn faceFromAxis(self: Self, axis: usize) FaceIndex {
            return .{
                .axis = axis,
                .side = self.sides[axis] == .right,
            };
        }

        /// Returns a space corresponding to the size of the region.
        pub fn space(self: Self, comptime E: usize, block: [N]usize) IndexSpace {
            var size: [N]usize = undefined;

            for (0..N) |i| {
                if (self.sides[i] != .middle) {
                    size[i] = E;
                } else {
                    size[i] = block[i];
                }
            }

            return IndexSpace.fromSize(size);
        }

        const naxis = "0";

        pub fn toString(self: Self) [N][]const u8 {
            var result: [N][]const u8 = undefined;

            for (0..N) |axis| {
                result[axis] = switch (self.sides[axis]) {
                    .left => (FaceIndex{ .axis = axis, .side = false }).toString(),
                    .right => (FaceIndex{ .axis = axis, .side = true }).toString(),
                    .middle => naxis,
                };
            }

            return result;
        }

        /// A type used for iterating nodes in a region.
        pub const NodeIterator = struct {
            inner: IndexSpace.CartesianIterator,
            block: [N]usize,
            sides: [N]Side,

            /// Increments to the next node.
            pub fn next(self: *NodeIterator) ?[N]isize {
                if (self.inner.next()) |cart| {
                    var result: [N]isize = undefined;

                    for (0..N) |i| {
                        switch (self.sides[i]) {
                            .left => result[i] = -1 - @as(isize, @intCast(cart[i])),
                            .right => result[i] = @intCast(self.block[i] + cart[i]),
                            else => result[i] = @intCast(cart[i]),
                        }
                    }

                    return result;
                } else {
                    return null;
                }
            }
        };

        /// Iterates all ghost nodes in this region out to a given extent `E`.
        pub fn nodes(self: Self, comptime E: usize, block: [N]usize) NodeIterator {
            return .{
                .inner = self.space(E, block).cartesianIndices(),
                .block = block,
                .sides = self.sides,
            };
        }

        /// A type used for interating nodes on the inner face of a region.
        pub const InnerFaceIterator = struct {
            inner: IndexSpace.CartesianIterator,
            block: [N]usize,
            sides: [N]Side,

            pub fn next(self: *InnerFaceIterator) ?[N]usize {
                if (self.inner.next()) |cart| {
                    var result: [N]usize = undefined;

                    for (0..N) |i| {
                        switch (self.sides[i]) {
                            .left => result[i] = 0,
                            .right => result[i] = self.block[i] - 1,
                            else => result[i] = cart[i],
                        }
                    }

                    return result;
                } else {
                    return null;
                }
            }
        };

        /// Iterates over all cells on the inner face.
        pub fn innerFaceCells(self: Self, block: [N]usize) InnerFaceIterator {
            var size: [N]usize = undefined;

            for (0..N) |i| {
                if (self.sides[i] == .left or self.sides[i] == .right) {
                    size[i] = 1;
                } else {
                    size[i] = block[i];
                }
            }

            return .{
                .inner = self.space(1, block).cartesianIndices(),
                .block = block,
                .sides = self.sides,
            };
        }

        pub const ExtentIterator = struct {
            inner: IndexSpace.CartesianIterator,
            sides: [N]Side,

            pub fn next(self: *ExtentIterator) ?[N]isize {
                if (self.inner.next()) |cart| {
                    var result: [N]isize = undefined;

                    for (0..N) |i| {
                        const off: isize = @intCast(cart[i]);

                        switch (self.sides[i]) {
                            .left => result[i] = -1 - off,
                            .right => result[i] = 1 + off,
                            else => result[i] = 0,
                        }
                    }

                    return result;
                } else {
                    return null;
                }
            }
        };

        /// Iterates outwards from the inner face. Provides offsets that can be added to the inner face
        /// index to find the global index.
        pub fn extentOffsets(self: Self, E: usize) ExtentIterator {
            // Determine size along each direction
            var size: [N]usize = undefined;

            for (0..N) |i| {
                if (self.sides[i] == .left or self.sides[i] == .right) {
                    size[i] = E;
                } else {
                    size[i] = 1;
                }
            }

            return .{
                .inner = IndexSpace.fromSize(size).cartesianIndices(),
                .sides = self.sides,
            };
        }

        pub fn extentDir(self: Self) [N]isize {
            var dir: [N]isize = undefined;

            for (0..N) |i| {
                switch (self.sides[i]) {
                    .left => dir[i] = -1,
                    .right => dir[i] = 1,
                    .middle => dir[i] = 0,
                }
            }

            return dir;
        }

        /// Returns a mask for which a given axis is set, if and only if `self.sides[axis] != .middle`.
        pub fn toMask(self: Self) AxisMask {
            var result = AxisMask.initEmpty();

            for (0..N) |axis| {
                result.setValue(axis, self.sides[axis] != .middle);
            }

            return result;
        }

        /// Returns a new region for which any axis such that `mask.isSet(axis) == false`,
        /// `self.sides[axis] == .middle`.
        pub fn masked(self: Self, mask: AxisMask) @This() {
            var sides: [N]Side = self.sides;

            for (0..N) |i| {
                if (!mask.isSet(i)) {
                    sides[i] = .middle;
                }
            }

            return .{
                .sides = sides,
            };
        }

        pub fn maskedBySplit(self: Self, split: AxisMask) @This() {
            var mask = AxisMask.initEmpty();

            for (0..N) |axis| {
                const side = self.sides[axis];
                switch (side) {
                    .middle => mask.unset(axis),
                    .left => mask.setValue(axis, split.isSet(axis) == false),
                    .right => mask.setValue(axis, split.isSet(axis)),
                }
            }

            return self.masked(mask);
        }

        // ************************
        // Constructors ***********
        // ************************

        /// Returns the null region.
        pub fn central() @This() {
            return .{
                .sides = [1]Side{.middle} ** N,
            };
        }

        /// Assembles an array of all valid regions.
        pub fn enumerate() [count]Region(N) {
            var regs: [count]Region(N) = undefined;

            const rspace = IndexSpace.fromSize([1]usize{3} ** N);

            var indices = rspace.cartesianIndices();

            var i: usize = 0;

            while (indices.next()) |cart| {
                var sides: [N]Side = undefined;

                for (0..N) |axis| {
                    sides[axis] = @enumFromInt(cart[axis]);
                }

                regs[i] = .{ .sides = sides };
                i += 1;
            }

            return regs;
        }

        /// Constructs an array of regions ordered by adjacency.
        pub fn enumerateOrdered() [count]Region(N) {
            var regs: [count]Region(N) = enumerate();
            insertion(Region(N), &regs, {}, lessThanFn);
            return regs;
        }

        /// Helper function for comparing adjacency.
        fn lessThanFn(_: void, lhs: Region(N), rhs: Region(N)) bool {
            return lhs.adjacency() < rhs.adjacency();
        }
    };
}

test "region adjacency" {
    const expectEqualSlices = std.testing.expectEqualSlices;

    const Side = Region(2).Side;

    const regions = Region(2).enumerateOrdered();

    try expectEqualSlices(Side, &regions[0].sides, &[_]Side{ .middle, .middle });
    try expectEqualSlices(Side, &regions[1].sides, &[_]Side{ .middle, .left });
    try expectEqualSlices(Side, &regions[2].sides, &[_]Side{ .left, .middle });
    try expectEqualSlices(Side, &regions[3].sides, &[_]Side{ .right, .middle });
    try expectEqualSlices(Side, &regions[4].sides, &[_]Side{ .middle, .right });
    try expectEqualSlices(Side, &regions[5].sides, &[_]Side{ .left, .left });
    try expectEqualSlices(Side, &regions[6].sides, &[_]Side{ .right, .left });
    try expectEqualSlices(Side, &regions[7].sides, &[_]Side{ .left, .right });
    try expectEqualSlices(Side, &regions[8].sides, &[_]Side{ .right, .right });
}

test "region indices" {
    const expect = std.testing.expect;
    const expectEqualSlices = std.testing.expectEqualSlices;

    const Side = Region(2).Side;

    const block: [2]usize = [_]usize{ 2, 2 };

    const region = Region(2){
        .sides = [_]Side{ .right, .left },
    };

    try expect(region.adjacency() == 2);

    var indices = region.nodes(2, block);

    try expectEqualSlices(isize, &[_]isize{ 2, -1 }, &indices.next().?);
    try expectEqualSlices(isize, &[_]isize{ 3, -1 }, &indices.next().?);
    try expectEqualSlices(isize, &[_]isize{ 2, -2 }, &indices.next().?);
    try expectEqualSlices(isize, &[_]isize{ 3, -2 }, &indices.next().?);
    try expect(indices.next() == null);

    var inner_cells = region.innerFaceCells(block);

    try expectEqualSlices(usize, &[_]usize{ 1, 0 }, &inner_cells.next().?);
    try expect(inner_cells.next() == null);

    var offsets = region.extentOffsets(2);

    try expectEqualSlices(isize, &[_]isize{ 1, -1 }, &offsets.next().?);
    try expectEqualSlices(isize, &[_]isize{ 2, -1 }, &offsets.next().?);
    try expectEqualSlices(isize, &[_]isize{ 1, -2 }, &offsets.next().?);
    try expectEqualSlices(isize, &[_]isize{ 2, -2 }, &offsets.next().?);
    try expect(offsets.next() == null);

    const other_region = Region(2){
        .sides = [_]Side{ .right, .right },
    };

    var other_offsets = other_region.extentOffsets(2);

    try expectEqualSlices(isize, &[_]isize{ 1, 1 }, &other_offsets.next().?);
    try expectEqualSlices(isize, &[_]isize{ 2, 1 }, &other_offsets.next().?);
    try expectEqualSlices(isize, &[_]isize{ 1, 2 }, &other_offsets.next().?);
    try expectEqualSlices(isize, &[_]isize{ 2, 2 }, &other_offsets.next().?);
    try expect(offsets.next() == null);
}

File: src/geometry/region.rs
use std::{
    array::from_fn,
    cmp::Ordering,
    fmt::{Display, Write},
};

use super::{index::IndexWindow, Split, CartesianIter, Face, IndexSpace};

/// Denotes where the region falls on a certain axis.
#[repr(u8)]
#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
pub enum Side {
    Left = 0,
    Middle = 1,
    Right = 2,
}

impl Side {
    pub fn reverse(self) -> Self {
        match self {
            Self::Left => Self::Right,
            Self::Right => Self::Left,
            Self::Middle => Self::Middle,
        }
    }

    pub fn from_value(val: u8) -> Self {
        assert!(val < 3);
        // Safety. We have specified the memory representation of the
        // enum and checked the value, so this should be safe.
        unsafe { std::mem::transmute(val) }
    }
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
pub struct Region<const N: usize> {
    #[serde(with = "crate::array")]
    sides: [Side; N],
}

impl<const N: usize> Region<N> {
    /// Number of different regions in a given number of dimensions.
    pub const COUNT: usize = 3usize.pow(N as u32);
    /// The default "central" region.
    pub const CENTRAL: Self = Self::new([Side::Middle; N]);

    // Builds a new region from the given sides.
    pub const fn new(sides: [Side; N]) -> Self {
        Self { sides }
    }

    pub const fn sides(&self) -> [Side; N] {
        self.sides
    }

    pub const fn side(&self, axis: usize) -> Side {
        self.sides[axis]
    }

    pub fn set_side(&mut self, axis: usize, side: Side) {
        self.sides[axis] = side
    }

    /// Reverse every side in the region.
    pub fn reverse(&self) -> Self {
        let mut result = [Side::Left; N];

        for axis in 0..N {
            result[axis] = self.sides[axis].reverse();
        }

        Self::new(result)
    }

    /// Returns number of axes that are not `Side::Middle`.
    pub fn adjacency(&self) -> usize {
        self.sides
            .into_iter()
            .filter(|&s| s != Side::Middle)
            .count()
    }

    /// Iterates over all faces one would have to move over to get to the region.
    pub fn adjacent_faces(&self) -> impl Iterator<Item = Face<N>> + '_ {
        (0..N)
            .filter(|&axis| self.side(axis) != Side::Middle)
            .map(|axis| Face {
                axis,
                side: self.side(axis) == Side::Right,
            })
    }

    /// Iterates over all splits that can touch this region.
    pub fn adjacent_splits(self) -> impl Iterator<Item = Split<N>> {
        let origin: [_; N] = from_fn(|axis| match self.side(axis) {
            Side::Left | Side::Middle => 0,
            Side::Right => 1,
        });

        let size: [_; N] = from_fn(|axis| match self.side(axis) {
            Side::Middle => 2,
            _ => 1,
        });

        IndexWindow::new(origin, size)
            .iter()
            .map(|index| Split::pack(from_fn(|axis| index[axis] != 0)))
    }

    /// Computes a split which touches the given region.
    pub fn adjacent_split(&self) -> Split<N> {
        let mut result = Split::empty();
        for axis in 0..N {
            result.set_to(axis, self.side(axis) == Side::Right)
        }
        result
    }

    /// Checks whether a given split is adjacent to the region.
    pub fn is_split_adjacent(&self, split: Split<N>) -> bool {
        for axis in 0..N {
            match (self.side(axis), split.is_set(axis)) {
                (Side::Left, true) => return false,
                (Side::Right, false) => return false,
                _ => {}
            }
        }

        true
    }

    // /// Returns an index space with the same size as the region.
    // pub fn index_space(&self, support: usize, block: [usize; N]) -> IndexSpace<N> {
    //     let mut size = [0; N];

    //     for axis in 0..N {
    //         if self.sides[axis] != Side::Middle {
    //             size[axis] = support;
    //         } else {
    //             size[axis] = block[axis]
    //         }
    //     }

    //     IndexSpace::new(size)
    // }

    // /// Iterates nodes in the given region (including ghost nodes and nodes
    // /// on faces).
    // pub fn nodes(&self, support: usize, block: [usize; N]) -> RegionNodeIter<N> {
    //     RegionNodeIter {
    //         inner: self.index_space(support, block).iter(),
    //         block,
    //         sides: self.sides,
    //     }
    // }

    // pub fn face_vertices(&self, block: [usize; N]) -> RegionFaceVertexIter<N> {
    //     let mut size = [0; N];

    //     for axis in 0..N {
    //         size[axis] = match self.sides[axis] {
    //             Side::Left | Side::Right => 1,
    //             Side::Middle => block[axis],
    //         }
    //     }

    //     RegionFaceVertexIter {
    //         inner: IndexSpace::new(size).iter(),
    //         block,
    //         sides: self.sides,
    //     }
    // }

    // pub fn offset_nodes(&self, support: usize) -> RegionOffsetNodeIter<N> {
    //     let size = self.sides.map(|side| match side {
    //         Side::Left | Side::Right => support,
    //         Side::Middle => 1,
    //     });

    //     RegionOffsetNodeIter {
    //         inner: IndexSpace::new(size).iter(),
    //         sides: self.sides,
    //     }
    // }

    // pub fn offset_dir(&self) -> [isize; N] {
    //     self.sides.map(|side| match side {
    //         Side::Left => -1,
    //         Side::Right => 1,
    //         Side::Middle => 0,
    //     })
    // }

    // /// Returns a mask for which a given axis is set if and only if `self.sides[axis] != Side::Middle`.
    // pub fn to_mask(&self) -> AxisMask<N> {
    //     let mut result = AxisMask::empty();

    //     for axis in 0..N {
    //         result.set_to(axis, self.sides[axis] != Side::Middle);
    //     }

    //     result
    // }

    // pub fn masked(&self, mask: AxisMask<N>) -> Self {
    //     let mut sides = self.sides;

    //     for i in 0..N {
    //         if !mask.is_set(i) {
    //             sides[i] = Side::Middle;
    //         }
    //     }

    //     Self::new(sides)
    // }

    // pub fn masked_by_split(&self, split: AxisMask<N>) -> Self {
    //     let mut mask = AxisMask::empty();

    //     for axis in 0..N {
    //         match self.sides[axis] {
    //             Side::Left => mask.set_to(axis, !split.is_set(axis)),
    //             Side::Middle => mask.clear(axis),
    //             Side::Right => mask.set_to(axis, split.is_set(axis)),
    //         }
    //     }

    //     self.masked(mask)
    // }

    /// Converts the region into an integer value.
    pub fn to_linear(&self) -> usize {
        let space = IndexSpace::new([3; N]);
        let index = from_fn(|axis| self.side(axis) as usize);
        space.linear_from_cartesian(index)
    }

    /// Converts an integer value into a region.
    pub fn from_linear(val: usize) -> Self {
        let space = IndexSpace::new([3; N]);
        let index = space.cartesian_from_linear(val);
        Self::new(from_fn(|axis| Side::from_value(index[axis] as u8)))
    }
}

impl<const N: usize> Ord for Region<N> {
    fn cmp(&self, other: &Self) -> Ordering {
        let space = IndexSpace::new([3; N]);
        let index = from_fn(|axis| self.side(axis) as usize);
        let oindex = from_fn(|axis| other.side(axis) as usize);

        space
            .linear_from_cartesian(index)
            .cmp(&space.linear_from_cartesian(oindex))
    }
}

impl<const N: usize> PartialOrd for Region<N> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl<const N: usize> Display for Region<N> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for axis in 0..N {
            match self.side(axis) {
                Side::Left => f.write_char('-'),
                Side::Middle => f.write_char('='),
                Side::Right => f.write_char('+'),
            }?;
        }
        Ok(())
    }
}

// /// Allows iterating the nodes in a region.
// pub struct RegionNodeIter<const N: usize> {
//     inner: CartesianIter<N>,
//     block: [usize; N],
//     sides: [Side; N],
// }

// impl<const N: usize> Iterator for RegionNodeIter<N> {
//     type Item = [isize; N];

//     fn next(&mut self) -> Option<Self::Item> {
//         let cart = self.inner.next()?;

//         let mut result = [0isize; N];

//         for axis in 0..N {
//             result[axis] = match self.sides[axis] {
//                 Side::Left => -(cart[axis] as isize),
//                 Side::Right => (self.block[axis] + cart[axis]) as isize,
//                 Side::Middle => cart[axis] as isize,
//             }
//         }

//         Some(result)
//     }
// }

// /// Allows iterating the vertices on the face of a region.
// pub struct RegionFaceVertexIter<const N: usize> {
//     inner: CartesianIter<N>,
//     block: [usize; N],
//     sides: [Side; N],
// }

// impl<const N: usize> Iterator for RegionFaceVertexIter<N> {
//     type Item = [usize; N];

//     fn next(&mut self) -> Option<Self::Item> {
//         let cart = self.inner.next()?;

//         let mut result = [0; N];

//         for axis in 0..N {
//             result[axis] = match self.sides[axis] {
//                 Side::Left => 0,
//                 Side::Right => self.block[axis] - 1,
//                 Side::Middle => cart[axis],
//             }
//         }

//         Some(result)
//     }
// }

// pub struct RegionOffsetNodeIter<const N: usize> {
//     inner: CartesianIter<N>,
//     sides: [Side; N],
// }

// impl<const N: usize> Iterator for RegionOffsetNodeIter<N> {
//     type Item = [isize; N];

//     fn next(&mut self) -> Option<Self::Item> {
//         let cart = self.inner.next()?;

//         let mut result = [0isize; N];

//         for axis in 0..N {
//             result[axis] = match self.sides[axis] {
//                 Side::Left => -(cart[axis] as isize),
//                 Side::Right => cart[axis] as isize,
//                 Side::Middle => 0,
//             }
//         }

//         Some(result)
//     }
// }

pub struct RegionIter<const N: usize> {
    inner: CartesianIter<N>,
}

impl<const N: usize> Iterator for RegionIter<N> {
    type Item = Region<N>;

    fn next(&mut self) -> Option<Self::Item> {
        Some(Region::new(self.inner.next()?.map(|idx| match idx {
            0 => Side::Left,
            1 => Side::Middle,
            2 => Side::Right,
            _ => unreachable!(),
        })))
    }
}

impl<const N: usize> ExactSizeIterator for RegionIter<N> {
    fn len(&self) -> usize {
        Region::<N>::COUNT
    }
}

/// Iterates over all regions in an N-dimensional space.
pub fn regions<const N: usize>() -> RegionIter<N> {
    RegionIter {
        inner: IndexSpace::new([3; N]).iter(),
    }
}

#[cfg(test)]
mod tests {
    use crate::geometry::{Split, Face};

    use super::{regions, Region, Side};

    #[test]
    fn region_iteration() {
        let comparison = [
            [Side::Left, Side::Left],
            [Side::Middle, Side::Left],
            [Side::Right, Side::Left],
            [Side::Left, Side::Middle],
            [Side::Middle, Side::Middle],
            [Side::Right, Side::Middle],
            [Side::Left, Side::Right],
            [Side::Middle, Side::Right],
            [Side::Right, Side::Right],
        ];

        for (region, compare) in regions().zip(comparison.into_iter()) {
            assert_eq!(region, Region::new(compare));
        }
    }

    #[test]
    fn adjacency() {
        let region = Region::new([Side::Left, Side::Right]);
        assert_eq!(region.adjacency(), 2);

        let mut faces = region.adjacent_faces();
        assert_eq!(faces.next(), Some(Face::negative(0)));
        assert_eq!(faces.next(), Some(Face::positive(1)));
        assert_eq!(faces.next(), None);

        let mut splits = region.adjacent_splits();
        assert_eq!(splits.next(), Some(Split::pack([false, true])));
        assert_eq!(splits.next(), None);

        assert_eq!(region.adjacent_split(), Split::pack([false, true]));
    }
}
